rfcs/007 - Peer review infra changes.md

```diff
@@ -0,0 +1,125 @@
+# 007 - Peer review infrastructure changes RFC
+
+### Proposed
+
+2022-12-16
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- David Ellis <david@iasql.com>
+- Yolanda Robla <yolanda@iasql.com>
+- Alejandro Guillen <alejandro@iasql.com>
+- Mohammad Pabandi <mohammad@iasql.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/iasql/iasql-engine/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Luis Fernando De Pombo <l@iasql.com>
+
+## Summary
+
+Eng teams move over from using cloud UIs to IaC because cloud infrastructure changes can lead to outages more often than business logic changes. As a result, it is considered a DevOps best practice to code review infrastructure changes using IaC tools and show the `preview` of the resulting change in pull requests to raise the quality bar and reduce the possibility of mistakes. Cloud infrastructure cannot be code reviewed with IaSQL using migration systems even if they support non-schema, or data, migrations (some ORMs like Prisma only support schema changes), and each migration is wrapped in an IaSQL transaction. Through our research and dogfood efforts, we have not found a migration system that fulfills the requirements needed to code review changes with IaSQL:
+- Allow data-only, `up` migrations
+- Allow plain SQL or have an introspective ORM
+- Free and ideally open source
+- Baseline/ignore certain migrations
+- Be resilient to `iasql_ugprade`
+
+Data migrations require being able to baseline migrations to avoid duplicate record violations on `iasql_upgrade`. Flyway comes closest to meeting all our requirements. However, it is not possible to baseline/ignore existing migrations on their free product.
+
+We could configure `iasql_upgrade` to preserve a `migrations` table and make the name of the table configurable, but the baselining issue would remain as `iasql_upgrade` can change the schema which renders existing migrations on past IaSQL versions invalid. As a result, we need an opinionated solution to unblock peer review of infra changes.
+
+## Proposal
```

Comment:
Thinking about it, a next step that can follow up this proposal could be adding the GitHub module, not to lock the process to a third-party vendor as GitHub but to demonstrate we have the ability to create this kind of module for third parties and can be integrated into the convention. For example, we can `get_sql_for_transaction` and then `gh_create_pr` or something and they can automate that step instead of adding the content and the preview directly from the IaSQL dashboard (without the approval process integration).

Reply:
Yeah, if it is automating the convention and not tightly integrated with the transaction workflow I think that is a great idea.